src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SharedScope.java

```diff
@@ -90,18 +88,30 @@ public HandleImpl acquire() {
                 throw new IllegalStateException("Already closed");
             } else if (value == MAX_FORKS) {
                 //overflow
-                throw new IllegalStateException("Segment acquire limit exceeded");
+                throw new IllegalStateException("Scope keep alive limit exceeded");
             }
         } while (!STATE.compareAndSet(this, value, value + 1));
-        return new SharedHandle();
+    }
+
+    @Override
+    @ForceInline
+    public void release0() {
+        int value;
+        do {
+            value = (int) STATE.getVolatile(jdk.internal.foreign.SharedScope.this);
```

Comment:
Doesn't need to be fully qualified I think?
```suggestion
            value = (int) STATE.getVolatile(this);
```